lib/snapshot/utils.go

```diff
@@ -319,3 +320,75 @@ func walkLinks(path, root string, linksWalked *int) (string, error) {
 		return filepath.Join(newdir, newpath), nil
 	}
 }
+
+// compare tar files of memFSNodes.
+func compareNode(node1, node2 *memFSNode, missing1, missing2 map[string]*memFSNode, diff1, diff2 map[string]*tar.Header, path string) {
+	if isSimilar, _ := tario.IsSimilarHeader(node1.hdr, node2.hdr, true); !isSimilar {
+		diff1[path] = node1.hdr
+		diff2[path] = node2.hdr
+	}
+
+	allChildren := make(map[string]bool)
+	for child := range node1.children {
+		allChildren[child] = true
+	}
+
+	for child := range node2.children {
+		allChildren[child] = true
+	}
+
+	for child := range allChildren {
+		nextNode1, ok1 := node1.children[child]
+		nextNode2, ok2 := node2.children[child]
+		updatedPath := filepath.Join(path, child)
+		if ok1 && ok2 {
+			compareNode(nextNode1, nextNode2, missing1, missing2, diff1, diff2, updatedPath)
+			continue
+		} else if ok1 {
+			missing2[updatedPath] = node1.children[child]
+		} else if ok2 {
+			missing1[updatedPath] = node2.children[child]
+		}
+	}
+}
+
+// log the info of a memFSNode.
+func logMemFSNodeInfo(node *memFSNode) {
+	hdr := node.hdr
+	log.Infof("%s %s %d %d %d", hdr.Name, hdr.FileInfo().Mode(), hdr.Uid, hdr.Gid, hdr.Size)
+	for _, nxtNode := range node.children {
+		logMemFSNodeInfo(nxtNode)
+	}
+}
+
+func CompareFS(fs1, fs2 *MemFS, image1Name, image2Name image.Name) {
+	missing1 := make(map[string]*memFSNode)
+	missing2 := make(map[string]*memFSNode)
+
+	diff1 := make(map[string]*tar.Header)
+	diff2 := make(map[string]*tar.Header)
+
+	compareNode(fs1.tree, fs2.tree, missing1, missing2, diff1, diff2, "")
+	image1Format := image1Name.GetRepository() + ":" + image1Name.GetTag()
+	image2Format := image2Name.GetRepository() + ":" + image2Name.GetTag()
+	// Files missing in first image but appeared in second image
+	log.Infof("======== file missing in first image %s=====", image1Format)
+	for _, node := range missing2 {
```

Comment:
should it be missing1 here?
I think we still need a test here.

Reply:
Nice catch!